py/torch_tensorrt/dynamo/conversion/impl/linear.py

```diff
@@ -0,0 +1,85 @@
+from typing import Optional, Union
+
+import tensorrt as trt
+import torch
+from torch.fx.node import Target
+from torch_tensorrt.fx.converters.converter_utils import (
+    SourceIR,
+    get_trt_tensor,
+    set_layer_name,
+    to_numpy,
+)
+from torch_tensorrt.fx.types import TRTNetwork, TRTTensor
+
+
+def linear(
+    network: TRTNetwork,
+    target: Union[Target, str],
+    source_ir: Optional[SourceIR],
+    name: str,
+    input: TRTTensor,
+    weight: Union[TRTTensor, torch.Tensor],
+    bias: Optional[Union[TRTTensor, torch.Tensor]],
+) -> TRTTensor:
+    """
+    TensorRT fully connected layer implicitly flatten last three dimensions at
+    the start and implicitly reshape the result to (K, 1, 1) at the end.
+
+    e.g. If input is (N, C, H, W), first it gets flatten to (N, C*H*W). Then after
+    going through fully connected operation, it becomes (N, K). Before sending it
+    out, it gets reshaped into (N, K, 1, 1) and this is the final output.
+
+    TODO: We can optimize this to get rid of unneccesary transformation.
+    """
+
+    if not isinstance(input, trt.tensorrt.ITensor):
+        raise RuntimeError(
+            f"Linear received input {input} that is not part " "of the TensorRT region!"
+        )
+
+    # reshape the input to (*, X, 1, 1)
+    pre_shuffle_layer = network.add_shuffle(input)
+    pre_shuffle_layer.reshape_dims = tuple(input.shape) + (1, 1)
+    set_layer_name(pre_shuffle_layer, target, f"{name}_pre_shuffle", source_ir)
+
+    # Process bias terms
+    if isinstance(bias, torch.Tensor):
+        # Transform the bias constant into a Numpy array
+        bias = to_numpy(bias)
+
+    elif isinstance(bias, TRTTensor):
+        bias = get_trt_tensor(network, bias, f"{name}_bias")
+
+    elif bias is not None:
+        raise RuntimeError(
+            f"Linear layer {name} has bias of type {type(bias)}, Expected Torch Tensor or TRT Tensor"
+        )
+
+    # Process weight terms
+    if network.has_explicit_precision or isinstance(weight, TRTTensor):
+        weight = get_trt_tensor(network, weight, f"{name}_weight")
+
+    elif isinstance(weight, torch.Tensor):
+        # Transform the weight constant into a Numpy array
+        weight = to_numpy(weight)
+
+    else:
+        raise RuntimeError(
+            f"Linear layer {name} has weight of type {type(weight)}, Expect Optional[Tensor]"
+        )
+
+    # add fully connected layer
+    fully_connected_layer = network.add_fully_connected(
```

Comment:
Do you mean `network.add_fully_connected(...)` is deprecated? I didn't find any prompt in the doc: https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Graph/Network.html#tensorrt.INetworkDefinition.add_fully_connected  Could you share where you found this?

Reply:
I think the layer deprecation can be found here: https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_fully_connected_layer.html. I do still think this converter is useful though. Potentially, it could invoke the `matrix_multiply`, after that PR gets merged?